### PR TITLE
[SPARK-28264][PYTHON][SQL] Support type hints in pandas UDF and rename/move inconsistent pandas UDF types

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to redesign pandas UDFs as described in [the proposal](https://docs.google.com/document/d/1-kV0FS_LF2zvaRh_GhkV32Uqksm_Sq8SvnBBmRyxm30/edit?usp=sharing).

```python
from pyspark.sql.functions import pandas_udf
import pandas as pd

@pandas_udf("long")
def plug_one(s: pd.Series) -> pd.Series:
    return s + 1

spark.range(10).select(plug_one("id")).show()
```

```
+------------+
|plug_one(id)|
+------------+
|           1|
|           2|
|           3|
|           4|
|           5|
|           6|
|           7|
|           8|
|           9|
|          10|
+------------+
```

Note that, this PR address one of the future improvements described [here](https://docs.google.com/document/d/1-kV0FS_LF2zvaRh_GhkV32Uqksm_Sq8SvnBBmRyxm30/edit#heading=h.h3ncjpk6ujqu), "A couple of less-intuitive pandas UDF types" (by @zero323) together.

In short,

- Adds new way with type hints as an alternative and experimental way.
    ```python
    @pandas_udf(schema='...')
    def func(c1: Series, c2: Series) -> DataFrame:
        pass
    ```

- Replace and/or add an alias for three types below from UDF, and make them as separate standalone APIs. So, `pandas_udf` is now consistent with regular `udf`s and other expressions.

    `df.mapInPandas(udf)`  -replace-> `df.mapInPandas(f, schema)`
    `df.groupby.apply(udf)`  -alias-> `df.groupby.applyInPandas(f, schema)`
    `df.groupby.cogroup.apply(udf)`  -replace-> `df.groupby.cogroup.applyInPandas(f, schema)`

    *`df.groupby.apply` was added from 2.3 while the other were added in the master only.

- No deprecation for the existing ways for now.
    ```python
    @pandas_udf(schema='...', functionType=PandasUDFType.SCALAR)
    def func(c1, c2):
        pass
    ```
If users are happy with this, I plan to deprecate the existing way and declare using type hints is not experimental anymore.

One design goal in this PR was that, avoid touching the internal (since we didn't deprecate the old ways for now), but supports type hints with a minimised changes only at the interface.

- Once we deprecate or remove the old ways, I think it requires another refactoring for the internal in the future. At the very least, we should rename internal pandas evaluation types.
- If users find this experimental type hints isn't quite helpful, we should simply revert the changes at the interface level.

### Why are the changes needed?

In order to address old design issues. Please see [the proposal](https://docs.google.com/document/d/1-kV0FS_LF2zvaRh_GhkV32Uqksm_Sq8SvnBBmRyxm30/edit?usp=sharing).

### Does this PR introduce any user-facing change?

For behaviour changes, No.

It adds new ways to use pandas UDFs by using type hints. See below.

**SCALAR**:

```python
@pandas_udf(schema='...')
def func(c1: Series, c2: DataFrame) -> Series:
    pass  # DataFrame represents a struct column
```

**SCALAR_ITER**:

```python
@pandas_udf(schema='...')
def func(iter: Iterator[Tuple[Series, DataFrame, ...]]) -> Iterator[Series]:
    pass  # Same as SCALAR but wrapped by Iterator
```

**GROUPED_AGG**:

```python
@pandas_udf(schema='...')
def func(c1: Series, c2: DataFrame) -> int:
    pass  # DataFrame represents a struct column
```

**GROUPED_MAP**:

This was added in Spark 2.3 as of SPARK-20396. As described above, it keeps the existing behaviour. Additionally, we now have a new alias `groupby.applyInPandas` for `groupby.apply`. See the example below:

```python
def func(pdf):
    return pdf

df.groupby("...").applyInPandas(func, schema=df.schema)
```

**MAP_ITER**: this is not a pandas UDF anymore

This was added in Spark 3.0 as of SPARK-28198; and this PR replaces the usages. See the example below:

```python
def func(iter):
    for df in iter:
        yield df

df.mapInPandas(func, df.schema)
```

**COGROUPED_MAP**: this is not a pandas UDF anymore

This was added in Spark 3.0 as of SPARK-27463; and this PR replaces the usages. See the example below:

```python
def asof_join(left, right):
    return pd.merge_asof(left, right, on="...", by="...")

 df1.groupby("...").cogroup(df2.groupby("...")).applyInPandas(asof_join, schema="...")
```

### How was this patch tested?

Unittests added and tested against Python 2.7, 3.6 and 3.7.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @rxin, @zero323, @cloud-fan, @mengxr, @viirya, @ueshin, @BryanCutler, @icexelloss, @dongjoon-hyun FYI

File: python/pyspark/sql/tests/test_pandas_cogrouped_map.py
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_cogrouped_map.py]: `groupby.cogroup.applyInPandas` now always sets the right eval type internally. So we don't have to worry about mis-typed UDF.

File: python/pyspark/sql/pandas/group_ops.py
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/group_ops.py]: I piggyback this change while I'm here.

File: python/pyspark/sql/pandas/functions.py
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/functions.py]: My logic here is that, we keep `PythonEvalType.SQL_COGROUPED_MAP_PANDAS_UDF` and `PythonEvalType.SQL_MAP_PANDAS_ITER_UDF` as internal flags for `mapInPandas` or `applyInPandas`; however, users don't set it directly since only `PandasUDFType` is the API

**Comment by HyukjinKwon** [comment]
For fixing examples and documentation, I will do it separately.

**Review by rberenguel — APPROVED** [review]

File: dev/sparktestsupport/modules.py
Comment (zero323) [inline dev/sparktestsupport/modules.py]: It looks like this contains only internal components. It wouldn't hurt to add leading underscore (probably applies to modules included with [SPARK-30434]).

Comment (HyukjinKwon) [reply dev/sparktestsupport/modules.py]: Let me keep it consistent in the current codebase for now. We could do that separately since there are already smilar private modules without underscore.

Comment (zero323) [reply dev/sparktestsupport/modules.py]: Makes sense, though we should probably keep that mind in the future. Latest refactoring of `pyspark.ml` moved things in that direction, so it would be nice to keep things consistent (maybe some official guidelines?).

**Review by zero323 — APPROVED** [review]

File: python/pyspark/sql/pandas/functions.py
Comment (viirya) [inline python/pyspark/sql/pandas/functions.py]: nit: Invalid function type? As like above `Invalid return type`?

File: python/pyspark/sql/pandas/group_ops.py
Comment (viirya) [inline python/pyspark/sql/pandas/group_ops.py]: Do we want to add some words that recommend to use `applyInPandas` now?

File: python/pyspark/sql/pandas/typehints.py
Comment (viirya) [inline python/pyspark/sql/pandas/typehints.py]: Do we need to make sure Ellipsis must at the end? Or it doesn't matter at all?

Comment (HyukjinKwon) [reply python/pyspark/sql/pandas/typehints.py]: I think it's always at the end:

```python
>>> import typing
>>> def func(a: typing.Tuple[str, ...]): pass
...
>>> def func(a: typing.Tuple[...]): pass
...
Traceback (most recent call last):
    ...
TypeError: Tuple[t0, t1, ...]: each t must be a type. Got Ellipsis.
>>> def func(a: typing.Tuple[..., str]): pass
...
Traceback (most recent call last):
    ...
TypeError: Tuple[t0, t1, ...]: each t must be a type. Got Ellipsis.
>>> def func(a: typing.Tuple[str, ..., int]): pass
...
Traceback (most recent call last):
    ...
TypeError: Tuple[t0, t1, ...]: each t must be a type. Got Ellipsis.
```

**Comment by viirya** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Should be ready for a look.

**Comment by BryanCutler** [comment]
This looks pretty good to me @HyukjinKwon , but I have a couple concerns

Using type hints can be a clear way to define the inputs/output of a UDF, but it might be problematic to be the only way a user can make a PandasUDF. What about if the user has a function they cannot modify to include type hints, or using lambda/partial functions? For these cases, it makes sense to use a wrapper like before, e.g. `pandas_udf(func)`. I'd also worry that forcing type hints might be off-putting to some users, since they are not that widely used or optional. I think it would be good to support a long-form PandasUDF declaration (like proposal 1) and also allow to infer the long-form from type hints (like this PR), so users have both options.

I'm also not too sure about changing some of the PandasUDFs to use regular functions, like with `df.groupby.apply(udf)`. I think it makes things less consistent by not using a `pandas_udf` for everything, and it could be inconvenient for the user to keep specifying the schema as an argument for multiple calls, instead of just binding it once with `pandas_udf`. It should be possible to still remove the udf type from the API so the following could be done which is almost similar to the proposed change:
```python
df.groupby.apply(pandas_udf(f, schema=...))
```

Anyway, I'm not opposed to anything here and fine with the new proposal, just some slight concerns before we commit to any new changes

**Comment by rxin** [comment]
On the first one they can wrap an existing function can’t they? They don’t
need to modify it.

**Comment by HyukjinKwon** [comment]
Yes, I think they can wrap it and should not be a big problem.

> I'd also worry that forcing type hints might be off-putting to some users, since they are not that widely used or optional.

That's my concern too which makes me to decide not to deprecate existing pandas UDFs for now. If we add proposal 1 way too, then we will have 3 ways (old one, proposal 1, and proposal 2) which I think might be more confusing. So my current take is to add the type hint way first, and see if users like it or not in Spark 3 (without any deprecation for now)

> I'm also not too sure about changing some of the PandasUDFs to use regular functions, like with df.groupby.apply(udf). I think it makes things less consistent by not using a pandas_udf for everything, and it could be inconvenient for the user to keep specifying the schema as an argument for multiple calls, instead of just binding it once with pandas_udf. It should be possible to still remove the udf type from the API so the following could be done which is almost similar to the proposed change:
>
> ```python
> df.groupby.apply(pandas_udf(f, schema=...))
> ```

I didn't have a strong preference on this change before. This was actually suggested by @zero323.
I took a look several times more and was convinced by my suggestion.

The current `udf` returns a `Column` which is able to bind with other SQL expressions whereas `pandas_udf` has three inconsistent cases `df.mapInPandas(udf)`, `df.groupby.apply(udf)` and `df.groupby.cogroup.apply(udf)`. Other expressions cannot be accepted in these APIs either (and looks impossible to fix it to take other expressions).

- If we can remove the three cases above, `pandas_udf` and `udf` are consistent with other functions as well.

- These naming of APIs `df.groupby.apply` and `df.groupby.cogroup.apply` are not clear if they are pandas UDF specific.

- Although we don't have UDF in SparkR, there are similar API shape in SparkR, `dapply` and `gapply`. So, this isn't completely new in Spark.

    ```r
    dapply(df, function(rdf) { ... }, structType("gear double")))
    ```

    ```r
    gapply(df, "k", function(k, g) { ... }, structType("gear double, disp boolean")))
    ```

**Comment by zero323** [comment]
@BryanCutler 

> Using type hints can be a clear way to define the inputs/output of a UDF, but it might be problematic to be the only way a user can make a PandasUDF. What about if the user has a function they cannot modify to include type hints, or using lambda/partial functions? For these cases, it makes sense to use a wrapper like before, e.g. pandas_udf(func).

While I agree with the conclusion (but there is no immediate plan to deprecate current style anyway) and had similar concerns these are mostly unjustified. Since suggested changes don't depend on static analysis, user is free to provide annotations on runtime:

```python
from pyspark.sql.pandas.typehints import infer_eval_type
from inspect import signature

def f(x: pd.Series, y: pd.Series) -> pd.Series:
    return x + y

infer_eval_type(signature(f))
## 200

def g(x, y):
    return x + y

infer_eval_type(signature(g))
## ValueError 
## ...
## ValueError: Type hints for all parameters should be specified; however, got (x, y)

g.__annotations__ =  {
    'x': pd.Series,
     'y': pd.Series,
    'return': pd.Series
}

infer_eval_type(signature(g))
## 200
```

Same approach can be used with `lambdas`, and in-lined with a simple wrapper, like this

```python
def with_annotations(f, annotations):
    f.__annotations = annotations
    return annotations
```

though given Pandas API design, lambdas are hardly useful.

As of partial functions - I am aware of more than one popular implementation, but if we're talking about `functools.partial`, `infer_eval_type` could be easily adjusted to eliminate `functools.partial.args` and `functools.partial.keywords` before proceeding.

>  I'd also worry that forcing type hints might be off-putting to some users, since they are not that widely used or optional. 

This proposal hardly forces usage of type hints, and in fact, it makes usage of type hints harder. It should be rather seen as a syntactic sugar, utilizing long established mechanism. 

Personally I don't have strong feelings about (concerns raised in the JIRA ticket and design document notwithstanding), but it should be pointed out, that mechanisms used here, are already utilized in the Python standard library, so cannot be seen as particularly exotic.

> I'm also not too sure about changing some of the PandasUDFs to use regular functions, like with df.groupby.apply(udf).

As mentioned by @HyukjinKwon I've agitated for this change and I don't want to repeat all the points I've raised before. I think it is important to address these two points:

> I think it makes things less consistent by not using a pandas_udf for everything, 

IMHO this is consistency from a perspective of a developer, not a user. It comes at significant price (growing number of execution types that user has to be aware of) and doesn't make any sense if you take a look at the types (please refer to design document for more detailed argument).

We do have methods that have similar semantics in both Scala  ("strongly" typed `Dataset` transformations), in SparkR (already mentioned) and 3rd party extensions (`sparklyr`'s  `spark_apply`). None requires end users to be aware of nitty-gritty details of the internal execution model.

It is a bit like saying that `KeyValueGroupedDataset,mapGroups` should be used as long as user provides corresponding `LogicalPlan` node. From where I stand forcing udf (with all the related complexity) is more a leaky abstraction, than a consistent design. 

> and it could be inconvenient for the user to keep specifying the schema as an argument for multiple calls,

For n-to-m, `pd.DataFrame` -> `pd.DataFrame` transformations users are as likely to have fixed schema, as schema varying based on the input Spark `DataFrame` (for example column additions and all-column aggregations). In such cases the alternative is to redefine `udf` for each possible input. Arguably this is at least as inconvenient and, in many cases hard to manage (as udf definitions have to be mixed with pipelines, to be able to access runtime schema).

File: python/pyspark/sql/pandas/typehints.py
Comment (zero323) [inline python/pyspark/sql/pandas/typehints.py]: Could we maybe

```
from typing import Union

all(
  issubclass(a, (pd.Series, pd.DataFrame)) 
  or a == Union[pd.Series, pd.DataFrame] 
  for a in parameters_sig
) and ...
```

to support cases where user doesn't know if input is atomic or struct?

**Comment by BryanCutler** [comment]
>On the first one they can wrap an existing function can’t they? They don’t
need to modify it.

Yeah, it's just an inconvenience for the user to define a wrapper function and then wrap that again in `pandas_udf`.

>That's my concern too which makes me to decide not to deprecate existing pandas UDFs for now. If we add proposal 1 way too, then we will have 3 ways (old one, proposal 1, and proposal 2) which I think might be more confusing. So my current take is to add the type hint way first, and see if users like it or not in Spark 3 (without any deprecation for now)

Sure, this sounds like a good path forward @HyukjinKwon . My concerns are only if type hints are required.

>The current udf returns a Column which is able to bind with other SQL expressions whereas 

Actually, `pandas_udf` returns a `UserDefinedFunction`. It's when that is called with `Columns` as input, it returns a `Column`.

>We do have methods that have similar semantics in both Scala ("strongly" typed Dataset transformations), in SparkR (already mentioned) and 3rd party extensions (sparklyr's spark_apply). None requires end users to be aware of nitty-gritty details of the internal execution model.

@zero323 Maybe I'm thinking of this from the users perspective, but to me `pandas_udf` in this context is really just a function bound with output schema. So it seems like we are just talking about 
```python
df.groupby("...").applyInPandas(func, schema=df.schema)
```
vs
```python
df.groupby("...").apply(pandas_udf(func, returnType=df.schema))
```
Which is just a subtle difference, so I'm not sure how this requires the user to the the details of internal execution.

**Comment by zero323** [comment]
> @zero323 Maybe I'm thinking of this from the users perspective, but to me `pandas_udf` in this context is really just a function bound with output schema. So it seems like we are just talking about
> 
> ```python
> df.groupby("...").applyInPandas(func, schema=df.schema)
> ```
> 
> vs
> 
> ```python
> df.groupby("...").apply(pandas_udf(func, returnType=df.schema))
> ```

Or some variant of that, where schema is derived from `df.schema`.

> 
> Which is just a subtle difference, so I'm not sure how this requires the user to the the details of internal execution.

From where I stand things look like this:

- As long as hints are not obligatory (and as mentioned in the design doc, types alone might not be sufficient to encode cardinality ) the latter will also require a `functionType`. This also affects which function types have to be exposed in the public API.

- This also informs user that internally use an UDF for execution and requires creation of an object, that is not suitable for use an UDF if we stick to public API. 

   And without looking at the function type (which alone is not friendly for interactive introspection) it is not exactly obvious, that object marked as UDF, and implementing UDF interface:

    ```
    class UDFLike(Protocol):
        def __call__(self, *__args: Column) -> Column: ....
    ```
    
    is not indented to be used as UDF at all. Looks rather confusing to me, but I admit I am rather used to looking at PySpark API in terms of types.

- Last but not least `pandas_udf` documentation  is already hard to digest - ~400 lines and growing.

The first and the last point affect anyone that uses `pandas_udf` (and the first one seems to drive discussion about API changes, so taking away some of the external complexity is good), even if user doesn't utilize this particular variants of UDF and I honestly cannot see any good reason for that ‒ PySpark users are used to higher order functions and getting wrapped function doesn't help to process data internally.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
gentle ping for a review .. :-)..

**Review by zero323 — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
I will probably merge this one in some days given the positive feedback from some people used to Python in the community. But I would love to get some sign-off from committers too honestly :-).

File: python/pyspark/sql/udf.py
Comment (BryanCutler) [inline python/pyspark/sql/udf.py]: does this need a trailing space `or"` -> `or "`?

File: python/pyspark/sql/tests/test_pandas_cogrouped_map.py
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_pandas_cogrouped_map.py]: Isn't this still a valid test?

File: python/pyspark/sql/pandas/group_ops.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/group_ops.py]: minor but should be plural -> `pandas.DataFrame`s. I'm not sure the correct format though.. I see `pandas.DataFrame`\\s and ``DataFrame``s below

File: python/pyspark/sql/pandas/functions.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/functions.py]: I'm not sure about this warning unless we are positive that we will be requiring type hints in the future. Plus, it seems to warn on `mapInPandas` etc, which isn't right..

**Review by BryanCutler — COMMENTED** [review]
I just had a few minor comments, but overall seems fine to me. Also, I think that `examples/sql/arrow.py` will need to be updated for `mapInPandas`, and cogroup.apply, but that could be done later.

**Comment by HyukjinKwon** [comment]
Thanks Bryan. I'll address them shortly.

File: python/pyspark/sql/pandas/functions.py
Comment (HyukjinKwon) [reply python/pyspark/sql/pandas/functions.py]: I actually would like to choose one side and encourage users to give feedbacks. Given my interactions here, using type hints seems preferred. Let me try to fix the condition.

**Comment by HyukjinKwon** [comment]
@BryanCutler, for fixing examples and documentation, I will do it separately in another PR.

File: python/pyspark/sql/pandas/group_ops.py
Comment (viirya) [inline python/pyspark/sql/pandas/group_ops.py]: Will `pyspark.sql.GroupedData.apply` be deprecated future? If so, and as it takes Python native func, will it be better not to say it is an alias of `apply`?

File: python/pyspark/sql/pandas/functions.py
Comment (viirya) [inline python/pyspark/sql/pandas/functions.py]: Shall we check `evalType` is `None` here?

File: python/pyspark/sql/tests/test_pandas_udf_typehints.py
Comment (viirya) [inline python/pyspark/sql/tests/test_pandas_udf_typehints.py]: Not sure about this. To confirm, does it mean we should not remove `exec` even after we drop Python 2 support?

**Review by viirya — APPROVED** [review]

**Review by BryanCutler — APPROVED** [review]

File: python/pyspark/sql/pandas/group_ops.py
Comment (ueshin) [inline python/pyspark/sql/pandas/group_ops.py]: Seems trying to always show this warning even when using `applyInPandas`?
Maybe we should move the following logic into `applyInPandas` and call it from here after showing the warning?

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM.

File: python/pyspark/sql/pandas/group_ops.py
Comment (HyukjinKwon) [reply python/pyspark/sql/pandas/group_ops.py]: I will do that when we actually deprecate. If users don't like this way, maybe we should remove this note and don't deprecate.

File: python/pyspark/sql/tests/test_pandas_udf_typehints.py
Comment (HyukjinKwon) [reply python/pyspark/sql/tests/test_pandas_udf_typehints.py]: It was a typo ...

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

Thanks you all for reviewing and participating in the discussion.

**Comment by zero323** [comment]
Side note ‒ I've been working lately with mentee whose code depends heavily on `pandas_udf` wrappers for third party functions, i.e.

```python
from foo import bar

pandas_udf(bar, schema, type)
```

one minor complain is that "will be deprecated" warnings tend to fill the log and notebook output (things are further complicated by the fact, that code is already annotated and reused, so injecting annotations is rather cumbersome).

So I guess there are three questions:

- Is the warning premature? Are plans to deprecate realistic, given current state of the ecosystem?
- Is warning applicable for explicit wrap, instead of decorator?
- Should we have more specific warning category than `UserWarning` that could be selectively suppressed?  Maybe something like `SparkDeprecationWarning` or `SparkPlannedDeprecationWarning`.

**Comment by HyukjinKwon** [comment]
Hey, sorry for my late response @zero323. I think I mistakenly missed this notification. I think warning is itself is fine as we didn't deprecate. I hope this warning encourages users to provide more feedback. I tried my best to collect feedback but still worry if we have enough feedback yet.

Better category, sure, yes. We should do that. We should actually throw a better exception too instead of just `Exception` in PySpark side.

Please let me know if you happen to have more feedback @zero323.

**Comment by zero323** [comment]
No worries @HyukjinKwon ‒ just trying to forward any feedback I get. I hope you're safe, wherever you are at the moment!

> Better category, sure, yes. We should do that. We should actually throw a better exception too instead of just Exception in PySpark side.

Yeah...

> Please let me know if you happen to have more feedback @zero323.

Sure thing, I will.

**Pull request closed without merging**
